### ISSUE TITLE
Test fails on MacOS with Intel MKL

### ISSUE THREAD

**Issue by theartful** [shogun-toolbox/shogun#4648]
I managed to reproduce the problem locally with gcc. It seems that using the Single Dynamic Library to build the tests causes problems. This is a sinppet that shows the problem:
```C++
#include <iostream>
#include <numeric>
#include <shogun/mathematics/Statistics.h>
#include <shogun/lib/SGMatrix.h>

int main()
{
	shogun::init_shogun();

	index_t num_samples = 100000;
	shogun::SGMatrix<float64_t> samples(2, num_samples);
	std::iota(samples.matrix, samples.matrix+num_samples, 1);
	
	auto emp_cov = shogun::CStatistics::covariance_matrix(samples, false);

	for (index_t i=0; i < emp_cov.num_rows*emp_cov.num_cols; ++i)
		std::cout << emp_cov.matrix[i] << '\n';
}
```
This piece of code when compiled using 
```
g++  main.cpp -o out -lshogun
```
or with
```
g++  main.cpp -o out -I/home/theartful/anaconda3/include/ -L/home/theartful/anaconda3/lib  -lshogun -lmkl_rt
```
produces different wrong results on each run. And it works fine when compiled using:
```
g++  main.cpp -o out -I/home/theartful/anaconda3/include/ -L/home/theartful/anaconda3/lib  -lshogun -lmkl_intel_lp64 -lmkl_gnu_thread -lmkl_core -lgomp -lpthread -lm -ldl
```
Replacing the single dynamic mkl library to build the tests seems to fix the problem.
I also tried to write code that uses Eigen using mkl. It works fine using `-lmkl_rt`, but if I also linked shogun `-lshogun` (without even using it in code) it behaves unpredictably, like the previous code sinppet.

**Comment by theartful** [comment]
Nvm what I said. The problem was solved locally by compiling against Intel OpenMP, instead of GNU OpenMP. Apparently MKL and gomp don't work together well. https://stackoverflow.com/questions/25986091/telling-gcc-to-not-link-libgomp-so-it-links-libiomp5-instead

**Comment by geektoni** [comment]
Fixed by #4656.
